### DISCUSSION TITLE
Constructing an independent ReadOnlyBase

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#2416] [Questions]
As we generally need to construct objects via the `DataPortal` rather than constructors, for a `ReadOnlyBase` which is independent and not part of a `ReadOnlyBindingListBase` which `DataPortal` method is the most appropriate? `DataPortal.FetchChild<T>(...)` is convenient because it accepts `params` thus being the most similar to a simple constructor with multiple parameters. Is this OK? Is there anything particularly special about `FetchChild` for read-only objects?

**Comment by rockfordlhotka** [comment]
I don't understand the scenario? 

You just want to create a readonly object without talking to the server? That's easily done by using `RunLocal`:

```c#
[Fetch]
[RunLocal]
private void Fetch(...)
{
}
```

Then

```c#
var obj = await DataPortal.FetchAsync<...>();
```

**Reply by michaelcsikos** [reply]
For `ReadOnlyBase` objects, is there any practical difference between `[Fetch]` and `[FetchChild]`?

**Reply by rockfordlhotka** [reply]
I guess it depends. 

In reality yes, there's a very big difference between the data portal pipeline for a root object as compared to a child object. There's a whole array of features around mobile objects, state management, transaction management, and more that exist only for root objects.

The child data portal does none of those things. It is merely an interceptor between the caller and callee to provide a superficial level of coding consistency for all types. Prior to the child data portal existing people had to manually craft their own solution for consistently managing child object persistence.

Now all that said, if you are willing to accept certain risks and limitations then maybe you are ok.

1. What you plan to do is not a supported concept, and there's no promise it would continue to work in the future
2. You do totally give up the ability to use a remote data portal, transaction management, and all the other features of the data portal

I'm not sure I see the point of assuming this risk. The calling code is basically the same, and the data portal operation method is also basically the same. So why assume the risk?

**Reply by michaelcsikos** [reply]
Just to clarify the scenario, with a typical `ReadOnlyBindingListBase` of `ReadOnlyBase` info objects, we do a `DataPortal.FetchChild` passing in a DTO, EF object, or could even just be some parameter values.

Sometimes we need to fetch just one `ReadOnlyBase` info object directly. As it's not contained in a `ReadOnlyBindingListBase` I just wasn't sure if I should being using `DataPortal.Fetch` or `DataPortal.FetchChild`.

In the past we've just used an `internal` constructor and created the instance directly.

**Reply by michaelcsikos** [reply]
Is there anything wrong with using an internal constructor for read-only objects?

**Reply by michaelcsikos** [reply]
``` cs
[Serializable]
public class ExampleInfo
    : ReadOnlyBase<ExampleInfo>
{
    public static readonly PropertyInfo<Guid> IdProperty = RegisterProperty<Guid>(nameof(Id));
    public Guid Id
    {
        get { return ReadProperty(IdProperty); }
        private set { LoadProperty(IdProperty, value); }
    }

    public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
    public string Name
    {
        get { return ReadProperty(NameProperty); }
        private set { LoadProperty(NameProperty, value); }
    }

    public static readonly PropertyInfo<string> CategoryProperty = RegisterProperty<string>(nameof(Category));
    public string Category
    {
        get { return ReadProperty(CategoryProperty); }
        private set { LoadProperty(CategoryProperty, value); }
    }

    // Is this OK?
    internal ExampleInfo(Guid id, string name, string category)
    {
        Id       = id;
        Name     = name;
        Category = category;
    }

    public ExampleInfo()
    {
    }
}
```

**Reply by rockfordlhotka** [reply]
The answer should be pretty straightforward.

If you are retrieving an object (list or single) from the client, then you should use the root methods (like `FetchAsync`).

If you are retrieving an object _from within another object while already on the server_ then you should use the child objects (like `FetchChildAsync`).

**Reply by rockfordlhotka** [reply]
btw, nothing stops your readonlybase type from having _both_ types of fetch operation method. 

```c#
  [Fetch]
  private void Fetch()
  {}

  [FetchChild]
  private void FetchChild()
  {}
```

And if they do the same thing (and require the same parameters), they can be the same method.

```c#
  [Fetch]
  [FetchChild]
  private void Fetch()
  {}
```
